### PR TITLE
statistics: split lock stats handler and rewrite the insert SQL

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/46351

Problem Summary:

### What is changed and how it works?

- Split the lock stats handler.
- Split the `AddLockedTables` function to small functions.
- Rewrote the INSERT SQL.
- Only try to execute the SQL if the table is not locked.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hi-rustin** [comment]
1. Start the TiDB cluster: `tiup playground nightly --tiflash 0 --db.binpath /Users/hi-rustin/GolandProjects/tidb/bin/tidb-server`
2. Create table and analyze it:
```sql
create table t(a int, b int);
insert into t values (1,2), (3,4), (5,6), (7,8);
analyze table t;
show warnings;
```
3. Lock table: `lock stats t;`
4. Analyze it again: `analyze table t;`
5. Check the warnings: `show warnings;`
```sql
```

File: statistics/handle/lock_stats_handler.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/statistics/handle/cache"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+// AddLockedTables add locked tables id  to store
```
Comment (qw4990) [inline statistics/handle/lock_stats_handler.go:31]:
```suggestion
// AddLockedTables add locked tables id to store
```

File: statistics/handle/lock_stats_handler.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+
+	_, err := exec.ExecuteInternal(ctx, "BEGIN PESSIMISTIC")
+	if err != nil {
+		return "", err
+	}
+
+	// Load tables to check duplicate when inserting.
+	rows, _, err := h.execRestrictedSQL(ctx, "SELECT table_id FROM mysql.stats_table_locked")
```
Comment (qw4990) [inline statistics/handle/lock_stats_handler.go:45]: It's a little weird here... First, we use `exec.Execute` to begin a new transaction, and then use another API `h.exec` to run a query. I don't know whether this is on purpose.

Comment (qw4990) [reply statistics/handle/lock_stats_handler.go:45]: If we change `h.exec` to `exec.Execute`, will it cause any failure? I just looked at some other functions like `SaveTableStatsToStorage`, it seems like we should always use `exec.Execute` here.

Comment (hi-rustin) [reply statistics/handle/lock_stats_handler.go]: Updated in [`ccb67d3` (#46403)](https://github.com/pingcap/tidb/pull/46403/commits/ccb67d3edde7aafd84996a8659ad677a479e3c64)

**Review by qw4990 — APPROVED** [review]

File: statistics/handle/lock_stats_handler.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+	if err != nil {
+		return "", err
+	}
+	rows, err := sqlexec.DrainRecordSet(ctx, recordSet, maxChunkSize)
+	if err != nil {
+		return "", err
+	}
+
+	dupTables := make([]string, 0)
+	tableLocked := make([]int64, 0)
```
Comment (hawkingrei) [inline statistics/handle/lock_stats_handler.go:55]: We add length for it.

File: statistics/handle/lock_stats_handler.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	for _, row := range rows {
+		tableLocked = append(tableLocked, row.GetInt64(0))
+	}
+
+	log := logutil.BgLogger().With(zap.String("category", "stats"))
```
Comment (hawkingrei) [inline statistics/handle/lock_stats_handler.go:60]: We can create a global one for reusing it.

**Review by time-and-fate — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**